### ISSUE TITLE
ElasticSearch Conflict Error on place order

### ISSUE THREAD

**Issue by rain2o** [DivanteLtd/magento2-vsbridge-indexer#347]
I originally posted this as a question in the forum - https://forum.vuestorefront.io/t/elasticsearch-conflict-error-on-place-order-magento-2/1852

We have been seeing this issue periodically where sometimes the place order API call fails because of an ElasticSearch Reindexing error, which is performed after the order has been placed. Due to this error, customers are placing the order multiple times because they believe it did not go through, when in fact the order was created successfully, it just failed during the reindex.

I have been able to track down that it comes from `\Divante\VsbridgeIndexerCatalog\Model\Indexer\Product::execute()`, and the error that is received is as follows:

```json
"failures": [
        {
            "index": "vue_storefront_catalog_en_eu_1582821702",
            "type": "product",
            "id": "38",
            "cause": {
                "type": "version_conflict_engine_exception",
                "reason": "[product][38]: version conflict, current version [1790] is different than the one provided [1789]",
                "index_uuid": "GjBqge-KRoWFREzBFuYZiA",
                "shard": "2",
                "index": "vue_storefront_catalog_en_eu_1582821702"
            },
            "status": 409
        }
    ]
```

I haven't been able to find a way to consistently reproduce it, so I haven't been able to find out much more information about why this happens. It seems like it is trying to update ES too quickly and doesn't have the updated reference, but I don't understand why this happens only sometimes. My first thought was maybe too many concurrent orders being placed at the same time, but we were able to reproduce this issue at least a few times in our test environment, so it is much less likely to be that.

I have seen recommendations to add parameters to the update request like ?conflicts=proceed&refresh=wait_for. However, as this is going through the VSBridge module, I haven’t been able to locate a place to add these parameters, or any way to figure out a way around this error.

I'm not really sure how to debug this further without a consistent way to reproduce, but if anyone has any suggestions for further testing/debugging, or if more information is needed please let me know and I will try to figure out how to get it.

**Comment by afirlejczyk** [comment]
I would try to add parameters in `Divante\VsbridgeIndexerCore\Elasticsearch\Client` - `bulk` method
but in my ES version I don't have `conflicts` parameter.
```
"conflicts" is not a valid parameter. Allowed parameters are "_source", "_source_excludes", "_source_includes", "error_trace", "filter_path", "human", "opaqueId", "pipeline", "pretty", "refresh", "routing", "source", "timeout", "type", "wait_for_active_shards"
``` 
but it may depends on ES version...

**Comment by rain2o** [comment]
Yes, I also tried the `conflicts` parameter, but noticed that it is not a valid parameter for the `_bulk` action. I believe instead it has to be passed in the body, in individual updates within the bulk, because that parameter is specific for the `_update` API.

Thanks for the link to the VSF CookBook, I tried searching for references to this issue in the docs but couldn't find it.

So you suggestion is to find a way to either skip entries when this error is thrown, or retry after fetching the updated version? All through the Bulk action?

As a side note, from what I can tell this reindex is getting triggered from a Product save after an order is placed I think. From what I can tell, this is probably a stock update or something. However, I find this interesting, because we typically have Manage Stock disabled for most of our products, except for special circumstances, so I find it interesting that the product is getting saved and then re-indexed as a result of an order. I'm wondering if there is a way to ignore this update for products with Manage Stock disabled, or if I'm missing something else that is requiring this re-index.

**Comment by rain2o** [comment]
I tried doing some testing to add the `retry_on_conflict` parameter inside the bulk body params, but I noticed that the bulk is just passing requests using the `index` API instead of the `update` API, and index doesn't appear to have that parameter. I'm not sure why it's passing these updates as `index` instead of `update`, seems like it's triggering the `Divante\VsbridgeIndexerCore\Api\BulkRequestInterface::addDocument` instead of `updateDocument`, but I haven't figured out why, or if this is actually a problem or not.

**Comment by afirlejczyk** [comment]
`updateDocument` will be removed, it's only used to partially update particular document. and is only used by vsbridge_product_category indexer which will will be removed.

So I would use `addDocument`for 100%.

At this time, I don't remember why I choose to use `index` to insert documents in bulk in ES.
Maybe we can use "update" instead of "index" but at the moment I can't tell you.
Maybe in our case - pushing bulk updates to elasticsearch both options are OK.
